zboxcore/sdk/sdk.go

```diff
@@ -868,6 +869,34 @@ func GetAllocationsForClient(clientID string) ([]*Allocation, error) {
 	return allocations, nil
 }
 
+// Q : Why do you have "Forbid" and "Allow" version of every option ?
+//
+// A : In Go, you cannot tell if the variable is initialized or not, it must have a value. For booleans,
+// the default value if "false". So if I decided to leave just the "forbid" versions, where if forbidX = 1, X is forbidden and v.v., 
+// the "false" value of the fobidX option will mean that the option X is allowed, however, it may be just that the user didn't even
+// set a value for it, so there's confusion between if the user has deliberaly set "false" to forbidX to "allow" the X option v.s. if
+// the user didn't give any value at all. Now for the "create" request, this confusion will not harm, since the two confused cases will
+// lead to the same result which is to "allow" the option X. However, for the "update" request, this confusion will lead to corrupted results.
+// Lets say the user wants to allow the option X in an update request, if there's only the "forbid" version then the user will send "forbidX=false".
+// However, since the user didn't touch the other options in their request, they're all considered "false", thus ALLOWED, even if another option Y
+// was dissallowed in the "create" request. That's why it's important to have a separate "allow" option so if the user wants to allow X they should
+// just send "allowX=true" which will affect the state of "X" only, where the others will stay unchanged.
+// In a nutshell, the idea is always using "true" as a sign of user's input and "false" as a sign of being unchanged or unset by the user.
```

Comment:
What you say is 100% correct (least that they should be allowed by default as mentioned by @guruhubb [here](https://github.com/0chain/0chain/issues/828#issuecomment-1003747207)), that's why for `newallocation` command in zboxcli I only added the `forbid_x` flags. The problem emerges when the owner needs to update the allocation and allow an action after he forbidden some actions. Let's look at this scenario if we're only using the `forbid` flags.
- Let's say originally both `delete` and `move` was forbidden.
- To allow delete, the user will need to run the following command `zbox updateallocation --allocation_id XXXXX --forbid_delete false`.
- Running this command the evaluation of the `forbid_delete` command is false, but also all the other flags including `forbid_move` are false as well, that's bcz `false` is Go's default for boolean values.

That's why I used both `allow` and `forbid` flags for update, to be able to find out if the user changed the flag explicitly or it's just a default. I'll re-check to see If I'll be able to do it without this double-flagging approach.

BTW same scenario as above emerges if we used `allow` flags alone. If all of them are allowed originally and the user runs an `updateallocation` command, even not touching any of the file options, they'll end up all forbidden as all the `allow` flags will parse to false by default.

Also If I were to use `flags.Changed`, I'll need to know there state before the command in `zboxcli` meaning I'll need to read the allocation in the command handling code in `zboxcli`, which is already read in gosdk so it will be read twice. I'll try to find a clean way to pass the `changed` state of the flags to gosdk.

Reply:
Resolved by passing flag change state to gosdk.

Reply:
It's not clear to me, why we can't do the same as linux does, when we update we give exact permissions, not the change in permissions

Reply:
Resolved as per DM

Reply:
@devyetii thanks for the details about the cases. My main point is to use `and` to unset the bits to disable actions. For example, suppose we have all enabled by default with value: `0011,1111`, and when we want to disable `upload` and `update`,  then we can simply do `value & ^(UploadMask+UpdateMask)`. To enable actions, suppose we have the `upload` and `update` disabled at the moment with value`0011,1010`, we can use `or` to enable them by `value | (UploadMask+UpdateMask)`. 

The above are the core code that will give the final file options value that will be used to update the allocation or file.  For the SDK API, we need to make it easy to use. That means we should only give the fields that we want to change rather than set all the fields, and it could be achieved easily with optional params. For example.
```go
// Define the file model option
type FileOption struct {
    BitMask int16
    Disable bool
}

func EnableFileOption(bitMask int16) FileOption {
    return FileOption{bm, false}
}

func DisableFileOption(bitMask int16) FileOption {
    return. File
}

func UpdateAllocation(xxxx, fileOpts ...FileOption) {
    // get init file mask
    fileOpt := alloc.FileOption
    // apply the options
    for _, fo := range fileOpts {
        if fo.Disable {
            fileOpt = fileOpt & ^(fo.BitMask)
        } else {
            fileOpt = fileOpt | fo.BitMask
        }
    }
}

```

Then we can only give the fields we want to update. 
For example, to disable `upload`, `update` and enable `rename`, `delete` by calling
```go
UpdateAllocation(alloc, xxxx, DisableFileOption(CanUploadMask|CanUpdateMask)), EnableFileOption(CanRenameMask|CanDeleteMask))
```

Reply:
Thanks for the review, Peter. I refactored the code to something like what you're saying.